src/main/frontend/handler/page.cljs

```diff
@@ -265,47 +265,80 @@
+         (when-let [file (d/entity (d/db conn) [:file/path old-path])]
+           (d/transact! conn  [{:db/id (:db/id file)
+                                :file/path new-path}])))
+
+       (p/let [_ (git/rename repo old-path new-path)]
+         (common-handler/check-changed-files-status)
+         (ok-handler)))
+     (p/catch (fn [error]
+                (println "file rename failed: " error))))))
+
 (defn rename!
   [old-name new-name]
   (when (and old-name new-name
              (not= (string/lower-case old-name) (string/lower-case new-name)))
     (when-let [repo (state/get-current-repo)]
-      (when-let [page (db/entity [:page/name (string/lower-case old-name)])]
-        (let [old-original-name (:page/original-name page)
-              file (:page/file page)]
-          (d/transact! (db/get-conn repo false)
-                       [{:db/id (:db/id page)
-                         :page/name (string/lower-case new-name)
-                         :page/original-name new-name}])
+      (if (db/entity [:page/name (string/lower-case new-name)])
+        (notification/show! "Page already exists!" :error)
+        (when-let [page (db/entity [:page/name (string/lower-case old-name)])]
+          (let [old-original-name (:page/original-name page)
+                file (:page/file page)]
+            (d/transact! (db/get-conn repo false)
+                         [{:db/id (:db/id page)
+                           :page/name (string/lower-case new-name)
+                           :page/original-name new-name}])
 
-          (when file
-            (page-add-properties! (string/lower-case new-name) {:title new-name}))
+            (when file
+              (rename-file! file new-name
+                            (fn []
+                              (page-add-properties! (string/lower-case new-name) {:title new-name}))))
 
-          ;; update all files which have references to this page
-          (let [files (db/get-files-that-referenced-page (:db/id page))]
-            (doseq [file-path files]
-              (let [file-content (db/get-file file-path)
-                    ;; FIXME: not safe
-                    new-content (string/replace file-content
-                                                (util/format "[[%s]]" old-original-name)
-                                                (util/format "[[%s]]" new-name))]
-                (file-handler/alter-file repo
-                                         file-path
-                                         new-content
-                                         {:reset? true
-                                          :re-render-root? false})))))
+            ;; update all files which have references to this page
+            (let [files (db/get-files-that-referenced-page (:db/id page))]
+              (doseq [file-path files]
+                (let [file-content (db/get-file file-path)
+                      ;; FIXME: not safe
+                      new-content (string/replace file-content
```

Comment:
Does it replace texts in code block?

Reply:
Not yet, that's why it's not safe.

Reply:
@gyk Do you want to fix this? I think it needs to be familiar with regex while I'm not :)

Reply:
This is probably difficult to solve by regex. You might have to parse the file to AST first, but it would slow down the code.

Reply:
@gyk We already have the ast in the database, it should be easy to walk through this ast to search and replace, but then the tricky part is that we need a way to generate a string using this modified ast which I think is quite hard to make it right.

Reply:
@tiensonqin Yes.